File: 116.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        vector<vector<TreeLinkNode*>> vec;
        bfs(root, 1, vec);
        for(int i=0; i<vec.size(); i++){
            for(int j=0; j<vec[i].size(); j++){
                if(j==vec[i].size()-1){
                    vec[i][j]->next == NULL;
                }else{
                    vec[i][j]->next = vec[i][j+1];
                }
            }
        }
    }
    
    void bfs(TreeLinkNode *root, int level, vector<vector<TreeLinkNode*>> &vec){
        if(!root) 
            return;
        if(vec.size()<level){
            vec.push_back(vector<TreeLinkNode*>{});
        }
        vec[level-1].push_back(root);
        bfs(root->left, level+1, vec);
        bfs(root->right, level+1, vec);
    }
};

File: golang/116/116.go
package main

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

func connect(root *Node) *Node {
	if root == nil {
		return root
	}

	queue := []*Node{root}

	for len(queue) > 0 {
		tmp := queue
		queue = make([]*Node, 0)

		for i, v := range tmp {
			if i+1 < len(tmp) {
				v.Next = tmp[i+1]
			}

			if v.Left != nil {
				queue = append(queue, v.Left)
			}

			if v.Right != nil {
				queue = append(queue, v.Right)
			}
		}
	}

	return root
}

/*
func connect(root *Node) *Node {
	if root == nil {
		return root
	}
	connTwoNode(root.Left, root.Right)
	return root
}

func connTwoNode(n1, n2 *Node) {
	if n1 == nil || n2 == nil {
		return
	}
	n1.Next = n2

	connTwoNode(n1.Left, n1.Right)
	connTwoNode(n2.Left, n2.Right)
	connTwoNode(n1.Right, n2.Left)
}
*/
